### PR TITLE
Added alignment of headers according to spec

### PR DESCRIPTION
**Pull request by AsafMah**
Added headers to kscb, tests, etc

### CODE REVIEW

File: kusto/conn.go [hunk 1]
```diff
@@ -8,6 +8,7 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
+	"github.com/google/uuid"
```
Comment (yogilad) [inline kusto/conn.go:11]: odd auto-align decision

File: kusto/conn.go [hunk 2]
```diff
@@ -39,10 +37,11 @@ type conn struct {
 	auth                           Authorization
 	endMgmt, endQuery, streamQuery *url.URL
 	client                         *http.Client
+	application, user, version     string
```
Comment (yogilad) [inline kusto/conn.go:40]:
```suggestion
	applicationForTracing, userForTracing, clientVersion     string
```

File: kusto/conn.go [hunk 3]
```diff
@@ -218,6 +207,30 @@ func (c *conn) doRequest(ctx context.Context, execType int, db string, query Stm
 	return op, header, resp.Header, body, nil
 }
 
+func (c *conn) getHeaders(properties requestProperties) http.Header {
+	header := http.Header{}
+	header.Add("Accept", "application/json")
+	header.Add("Accept-Encoding", "gzip")
+	header.Add("Content-Type", "application/json; charset=utf-8")
+	header.Add("x-ms-version", "2019-02-13")
+	header.Add("x-ms-client-request-id", "KGC.execute;"+uuid.New().String())
```
Comment (yogilad) [inline kusto/conn.go:216]: Don't we support user provided ClientRequestId in Go?

File: kusto/ingest/internal/conn/conn.go [hunk 4]
```diff
@@ -32,17 +31,18 @@ var BuffPool = sync.Pool{
...
-	baseURL    *url.URL
-	reqHeaders http.Header
-	client     *http.Client
-	done       chan struct{}
+	auth                       kusto.Authorization
+	baseURL                    *url.URL
+	reqHeaders                 http.Header
+	client                     *http.Client
+	done                       chan struct{}
+	application, user, version string
```
Comment (yogilad) [inline kusto/ingest/internal/conn/conn.go:39]: same as with Kusto/conn.go

File: kusto/ingest/internal/conn/conn.go [hunk 5]
```diff
@@ -51,15 +51,18 @@ func New(endpoint string, auth kusto.Authorization, client *http.Client) (*Conn,
 		).SetNoRetry()
 	}
 
-	return newWithoutValidation(endpoint, auth, client)
+	return newWithoutValidation(endpoint, auth, client, application, user, version)
 }
 
-func newWithoutValidation(endpoint string, auth kusto.Authorization, client *http.Client) (*Conn, error) {
+func newWithoutValidation(endpoint string, auth kusto.Authorization, client *http.Client, application, user, version string) (*Conn, error) {
```
Comment (yogilad) [inline kusto/ingest/internal/conn/conn.go:57]: Why is this a duplicate of Kusto/conn.go
Also, for streamingingestion you can pass CRP so this should look similar to the other conn.

File: kusto/ingest/streaming.go [hunk 6]
```diff
@@ -32,7 +32,7 @@ var FileIsBlobErr = errors.ES(errors.OpIngestStream, errors.KClientArgs, "blobst
 // More information can be found here:
 // https://docs.microsoft.com/en-us/azure/kusto/management/create-ingestion-mapping-command
 func NewStreaming(client QueryClient, db, table string) (*Streaming, error) {
-	streamConn, err := conn.New(client.Endpoint(), client.Auth(), client.HttpClient())
+	streamConn, err := conn.New(client.Endpoint(), client.Auth(), client.HttpClient(), "", "", "")
```
Comment (yogilad) [inline kusto/ingest/streaming.go:35]: Is not having CRPs here mapped as a parity gap, or is it the same in all SDKs?

File: kusto/kcsb.go [hunk 7]
```diff
@@ -28,9 +34,15 @@ type ConnectionStringBuilder struct {
 	RedirectURL                      string
 	DefaultAuth                      bool
 	ClientOptions                    *azcore.ClientOptions
+	ApplicationForTracing            string
+	UserForTracing                   string
+	versionForTracing                string
```
Comment (yogilad) [inline kusto/kcsb.go:39]: This is not a KCSB capability

File: kusto/kcsb.go [hunk 8]
```diff
@@ -163,6 +175,35 @@ func (kcsb *ConnectionStringBuilder) resetConnectionString() {
...
+func (kcsb *ConnectionStringBuilder) setDefaultTracingValues() {
+	kcsb.ApplicationForTracing = filepath.Base(os.Args[0])
+	kcsb.UserForTracing = getOsUser()
+	kcsb.versionForTracing = buildHeaderFormat(StringPair{Key: "Kusto.Go.Client", Value: version.Kusto}, StringPair{Key: "Go", Value: runtime.Version()})
+}
+
+func getOsUser() string {
+	var final string
+	current, err := user.Current()
+	if err != nil || current.Username == "" {
```
Comment (yogilad) [inline kusto/kcsb.go:191]: nit: this is odd cause you used the checked value in the else
Reverse the condition perhaps?
You can also consider scope reduction (if {return} X N)

File: kusto/ingest/streaming.go (hunk 6 above)
Comment (yogilad) [reply kusto/ingest/streaming.go:35]: Also, should we not take this from 'client'?

File: kusto/kusto.go [hunk 9]
```diff
@@ -381,6 +386,27 @@ func (c *Client) HttpClient() *http.Client {
 	return c.http
 }
 
+func (c *Client) Application() string {
+	if conn, ok := c.conn.(*conn); ok {
```
Comment (yogilad) [inline kusto/kusto.go:390]: not sure I understand.

File: kusto/conn.go (hunk 3 above)
Comment (AsafMah) [reply kusto/conn.go:216]: Nope, but I'll add

File: kusto/kcsb.go (hunk 7 above)
Comment (AsafMah) [reply kusto/kcsb.go:39]: WDYM? I'll move the logic to another class, but we still need getters/setters for these in KCSB

File: kusto/ingest/internal/conn/conn.go (hunk 5 above)
Comment (AsafMah) [reply kusto/ingest/internal/conn/conn.go:57]: conn.go is a private type, so we need to duplicate it here in this package.
Alternatively we can have this functionality be a part of the regular "conn" under an interface and reduce the duplication maybe

File: kusto/ingest/streaming.go (hunk 6 above)
Comment (AsafMah) [reply kusto/ingest/streaming.go:35]: No SDK lets you send CRP to streaming ingestion

Comment (AsafMah) [reply kusto/ingest/streaming.go]: In fact, the endpoint doesn't expect it either..

**Comment by AsafMah** [comment]
@yogilad I refactored it to remove the logic out of KSCB, take another look.

File: kusto/client_details.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	if isEmpty(final) {
+		final = NONE
+	}
+
+	return final
+}
+
+var escapeRegex = regexp.MustCompile("[\\r\\n\\s{}|]+")
```
Comment (yogilad) [inline kusto/client_details.go:66]: What's the Go best practice on this?
Should also consts be grouped some place or is it ok to spread them around?

**Review by yogilad — APPROVED** [review]

File: kusto/client_details.go (hunk 10 above)
Comment (AsafMah) [reply kusto/client_details.go:66]: I'll put it with the other consts on the top.

**Review by yogilad — APPROVED** [review]

**Pull request merged**